### PR TITLE
Add volume cloning test

### PR DESCRIPTION
**Pull request by khushboo-rancher**
https://github.com/longhorn/longhorn/issues/2941

This is implementation of tests in the volume cloning module..

I did below modifications in the test skeleton:
1. Renamed test case name.
2. Previously cloning from attached and detached volume was part one test, now divided them in two.

Signed-off-by: Khushboo <fnu.khushboo@suse.com>

### CODE REVIEW

**Review by chanow816 — DISMISSED** [review]

File: manager/integration/tests/common.py [hunk 1]
```diff
@@ -4467,6 +4475,31 @@ def wait_for_pod_annotation(core_api,
...
 
 
+def wait_for_volume_clone_status(client, name, key, value):
+    for _ in range(RETRY_COUNTS):
+        volume = client.by_id_volume(name)
+        if volume[VOLUME_FIELD_CLONE_STATUS][key] == value:
+            break
+        time.sleep(RETRY_INTERVAL)
+    assert volume[VOLUME_FIELD_CLONE_STATUS][key] == value, \
+        f" value={value}\n. volume[cloneStatus][key]=" \
```
Comment (PhanLe1010) [inline manager/integration/tests/common.py:4485]:
```suggestion
        f" value={value}\n. volume[{VOLUME_FIELD_CLONE_STATUS}][{key}]=" \
```

Comment (PhanLe1010) [inline manager/integration/tests/common.py:4485]:
```suggestion
        f" Expecting value={value}\n. Got volume[{VOLUME_FIELD_CLONE_STATUS}][{key}]=" \
```

File: manager/integration/tests/test_cloning.py [hunk 2]
```diff
@@ -36,49 +94,178 @@ def test_cloning_without_backing_image():  # NOQA
...
-    6. Wait for the `CloneStatus.State` in `cloned-pvc` to be `completed`
-    7. In 3-min retry loop, wait for the cloned pod to be running
-    8. Verify the data in `cloned-pvc` is the same as in `source-pvc`
-    9. In 2-min retry loop, verify the volume of the `clone-pvc` eventually
+    5. Wait for the `CloneStatus.State` in `cloned-pvc` to be `completed`
+    6. Clone volume should get detached after cloning completion, wait for it.
+    7. Specify the `cloned-pvc` in a cloned pod yaml and deploy the cloned pod
+    8. In 3-min retry loop, wait for the cloned pod to be running
+    9. Verify the data in `cloned-pvc` is the same as in `source-pvc`
+    10. In 2-min retry loop, verify the volume of the `clone-pvc` eventually
```
Comment (PhanLe1010) [inline manager/integration/tests/test_cloning.py:104]: Just want to double check, are we cleaning up the pods and the PVC at the end of the test?

File: manager/integration/tests/test_cloning.py [hunk 3]
```diff
@@ -36,49 +94,178 @@ def test_cloning_without_backing_image():  # NOQA
...
+                name: source-pvc
+                kind: PersistentVolumeClaim
+              accessModes:
+                - ReadWriteOnce
+              resources:
+                requests:
+                  storage: 10Gi
+            ```
+        6. Wait for `source-pvc` to be attached
+        7. Wait for a new snapshot created in `source-pvc` volume created
```
Comment (PhanLe1010) [inline manager/integration/tests/test_cloning.py:198]: If the cloning happens fast enough, we might not see these intermediate steps. Should we skip them?

File: manager/integration/tests/test_cloning.py [hunk 4]
```diff
@@ -140,14 +327,90 @@ def test_cloning_interrupted():  # NOQA
...
+
+    # Step-2
+    pod_name = 'source-pod' + generate_random_suffix()
+    pod['metadata']['name'] = pod_name
+    pod['spec']['volumes'] = [create_pvc_spec(source_pvc_name)]
+    create_and_wait_pod(core_api, pod)
+
+    # Step-3
+    write_pod_volume_random_data(core_api, pod_name,
+                                 '/data/test', DATA_SIZE_IN_MB_3)
```
Comment (PhanLe1010) [inline manager/integration/tests/test_cloning.py:356]: Can we modify the test description to specify that we only write 500MB of data?

**Review by PhanLe1010 — COMMENTED** [review]
In general LGTM.

Thank you for implementing the test 🚀

File: manager/integration/tests/common.py (hunk 1 above)
Comment (khushboo-rancher) [reply manager/integration/tests/common.py]: Fixed, thanks

File: manager/integration/tests/test_cloning.py (hunk 4 above)
Comment (khushboo-rancher) [reply manager/integration/tests/test_cloning.py:356]: Fixed, thanks

File: manager/integration/tests/test_cloning.py (hunk 2 above)
Comment (khushboo-rancher) [reply manager/integration/tests/test_cloning.py:104]: Yes, the pods and pvc are getting cleaned up in the fixtures.

File: manager/integration/tests/test_cloning.py (hunk 3 above)
Comment (khushboo-rancher) [reply manager/integration/tests/test_cloning.py:198]: I ran this multiple times, we get enough time to get these states and I think these are important to check. 
If encounter problem with these checks in the Jenkins job runs, we can skip these checks.

Comment (PhanLe1010) [reply manager/integration/tests/test_cloning.py:198]: Okay. Thank you for the explanation

**Review by PhanLe1010 — APPROVED** [review]
LGTM

File: manager/integration/tests/common.py [hunk 5]
```diff
@@ -4467,6 +4475,32 @@ def wait_for_pod_annotation(core_api,
...
+
+def get_clone_volume_name(client, source_volume_name):
+    for _ in range(RETRY_EXEC_COUNTS):
+        volumes = client.list_volume()
+        for volume in volumes:
+            if volume['cloneStatus']['sourceVolume'] == \
+                    source_volume_name:
+                return volume.name
+            else:
+                continue
```
Comment (innobead) [inline manager/integration/tests/common.py:4499]: nit: we don't need this else condition, because at the end of loop, it will just reiterate the next.

**Review by innobead — CHANGES_REQUESTED** [review]
LGTM, but just one minor comment needed to fix.

File: manager/integration/tests/common.py (hunk 5 above)
Comment (khushboo-rancher) [reply manager/integration/tests/common.py]: Right, we don't need it, removed it. :)

**Review by innobead — APPROVED** [review]

**Pull request merged**